cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,514 @@
+    return store.getAllAdapters(Id.Namespace.from(namespace));
+  }
+
+  /**
+   * Creates the adapter
+   *
+   * @param namespaceId namespace
+   * @param adapterSpec instance of {@link AdapterSpecification}
+   * @throws IllegalArgumentException
+   */
+  public void createAdapter(String namespaceId, AdapterSpecification adapterSpec) throws IllegalArgumentException {
+
+    AdapterTypeInfo adapterTypeInfo = adapterTypeInfos.get(adapterSpec.getType());
+    Preconditions.checkNotNull(adapterTypeInfo, "Adapter type %s not found", adapterSpec.getType());
+
+    try {
+      ApplicationSpecification appSpec = deployApplication(namespaceId, adapterTypeInfo);
+
+      validateSources(adapterSpec.getName(), adapterSpec.getSources());
+      createSinks(adapterSpec.getSinks(), adapterTypeInfo);
+
+      // If the adapter already exists, remove existing schedule to replace with the new one.
+      AdapterSpecification existingSpec = store.getAdapter(Id.Namespace.from(namespaceId), adapterSpec.getName());
+      if (existingSpec != null) {
+        stopPrograms(namespaceId, appSpec, adapterTypeInfo, existingSpec);
+      }
+
+      startPrograms(namespaceId, appSpec, adapterTypeInfo, adapterSpec);
+      store.addAdapter(Id.Namespace.from(namespaceId), adapterSpec);
+
+    } catch (Exception e) {
+      throw new RuntimeException("Error deploying adapter application " + e.getMessage());
+    }
+  }
+
+  /**
+   * Remove adapter identified by the namespace and name.
+   *
+   * @param namespace namespace id
+   * @param adapterName adapter name
+   * @throws AdapterNotFoundException if the adapter to be removed is not found.
+   */
+  public void removeAdapter(String namespace, String adapterName) throws AdapterNotFoundException {
+    Id.Namespace namespaceId = Id.Namespace.from(namespace);
+    AdapterSpecification adapterSpec = store.getAdapter(namespaceId, adapterName);
+    if (adapterSpec == null) {
+      throw new AdapterNotFoundException(String.format("Adapter %s not found.", adapterName));
+    }
+
+    ApplicationSpecification appSpec = store.getApplication(Id.Application.from(namespaceId, adapterSpec.getType()));
+    stopPrograms(namespace, appSpec, adapterTypeInfos.get(adapterSpec.getType()), adapterSpec);
+    store.removeAdapter(namespaceId, adapterName);
+
+    // TODO: Delete the application if this is the last adapter
+  }
+
+  public void stopAdapter(String namespace, String adapterName) {
+    //TODO:
+  }
+
+  public void startAdapter(String namespace, String adapterName) {
+    //TODO:
+  }
+
+  // Deploys adapter application if it is not already deployed.
+  private ApplicationSpecification deployApplication(String namespaceId, AdapterTypeInfo adapterTypeInfo)
+    throws Exception {
+
+    ApplicationSpecification spec = store.getApplication(Id.Application.from(namespaceId, adapterTypeInfo.getType()));
+    // Application is already deployed.
+    if (spec != null) {
+      return spec;
+    }
+
+    try {
```

Comment:
seems like there is no need for the try since we're just propagating the exception